src/sentry/search/moresnuba/backend.py

```diff
@@ -0,0 +1,149 @@
+from __future__ import absolute_import
+import logging
+from sentry.search.snuba import SnubaSearchBackend
+from sentry.utils import snuba
+
+
+# TODO:
+# Let's give this backend a self explanatory name and a docstring that explains why it exists.
+# Written this way it would be very hard for someone reading the code to understand the design.
+# This backend is intended to "extend" the current events search backend.
+# The goal of it is to transition to searches that use snuba more heavily - perhaps ideally not using postgres/django models at all
+
+# This file overrides certain methods, and variables in order to do this.
+# Feel free to break the base backend up more if you need to modify certain bits of functionality in here.
+# We alias field names with "groups." or "events." where appropriate, use a different dataset,
+# and remove some issue_only_fields and query building magic from the backup.
+
+# We are running this backend alongside the original one and logging results.
+# Eventually, we hope to remove the original backend and use this if it performs well.
+class MoreSnubaSearchBackend(SnubaSearchBackend):
```

Comment:
It's not that we're moving entirely away from postgres - we're just making less queries on it. For example, we won't be querying for group status, last_seen, first_seen, and active_at anymore. Eventually, I think the goal is to be able to query other fields (assigned_to, unassigned, subscribed_by, maybe others) from snuba as well and then we won't need to do any pre or post filtering via postgres. I think at that point we can remove the estimator/sampling logic, and much of the other stuff and rely entirely on snuba.

The "moresnuba" optimizations are still a WIP in this PR - it will just include changes (probably in another pr once I get the tests in) to not query on the group status field, last_seen, first_seen, and active_at in postgres. It will use the events or groups (depending on whether we have an environment) filter to get results on those fields.

Reply:
I looked into this as well, and I think it might have complexities.

The current system assumes pre or post filtering will be done. It gets a list of candidate_ids from postgres based on filters, and if this isn't greater than max_candidates, we pass it to snuba. If it's empty, we assume no results and return that with no snuba query. If is is greater than the max_candidates, then we do post filtering and hits estimation (which relies on the group queryset to calculate hit ratio).

My question is:
If we have no group queryset information, how would we estimate the hits? We'd need a new way - maybe using totals from Snuba? Or could we do what we do with candidate ids? I'm assuming not because results could be greater than the maximum snuba could return (1000?).

And the actual query loop in chunks leaves exits immediately with the results and hits count if we had the pre filter, but does some verifying against the queryset if we are post filtering. So, I'm wondering, if I would just remove this and just append to result groups with no verification against the queryset (just verification to ensure no duplicates). But again, the hits? Could they come from the totals? or is that also capped at 1000?

so I'm down to leave this optimization out for now.

Reply:
I would not call "removing the postgres query" an optimization. That's basically the point of using a joined dataset in Snuba, preserving that query defies the purpose of all this work. 
I would not go ahead without having a clear plan on how to skip that query.

About the specific cases of pre/post filtering
- post filtering. This will not exist on a Snuba-only query, since Snuba is able to do all the filtering
- calculating hits when candidate list is too big. This is not a case that exists in a Snuba only query because Snuba is doing the filtering
- calculating hits in case of a cursor. I think, in a Snuba-only query we should be able to apply the same approach we do today without filtering according to the postgres results since Snuba is already doing all the proper filtering.

So the post filtering should not be a problem. Clearly, running the query in Snuba only, imply differences in the way we do the query that go beyond changing the list of conditions. This is actually an additional argument for moving the whole query logic into a method that would be  implemented radically differently between postgres+snuba queries and snuba queries.